### ISSUE TITLE
Compilation error cygwin.

### ISSUE THREAD

**Issue by jun0s** [magnumripper/JohnTheRipper#1886]
/aes.a -o ../run/john
x86.o:fake:(.text+0x10af): referencia a `CPU_detect_pop_ecx_fail' undefined

**Comment by magnumripper** [comment]
Thanks for reporting. Please try 5662eaf and report results!

**Comment by jun0s** [comment]
check this...
$ ./john.exe
Sorry, SSSE3 is required for this build

**Comment by magnumripper** [comment]
I will investigate. Our CircleCI build bot failed too, with "Sorry, SSE2 is required for this build" even though it has SSE2.

**Comment by magnumripper** [comment]
I'm hoping for 968134a now :wink:

**Comment by jun0s** [comment]
It remains the same with the last change.. 

$ ./john.exe
Sorry, SSSE3 is required for this build

**Comment by magnumripper** [comment]
386e1b7 is verified to work. Please re-open if not.

**Comment by magnumripper** [comment]
It turns out it worked for ANY cpu except exactly SSE2 :laughing: 

Fixed in 258b18a

**Comment by magnumripper** [comment]
Doh. I will investigate.

**Comment by jun0s** [comment]
thank u!!!

**Comment by magnumripper** [comment]
30b660f
```
Benchmarking: NT [MD4 128/128 AVX 4x2]... DONE
Raw:	32528K c/s real, 32528K c/s virtual

Benchmarking: NT [MD4 128/128 SSE4.1 4x2]... DONE
Raw:	31001K c/s real, 31001K c/s virtual

Benchmarking: NT [MD4 128/128 SSSE3 4x2]... DONE
Raw:	31301K c/s real, 31301K c/s virtual

Benchmarking: NT [MD4 128/128 SSE2 4x2]... DONE
Raw:	30840K c/s real, 30840K c/s virtual

Benchmarking: NT [MD4 32/32]... DONE
Raw:	5361K c/s real, 5361K c/s virtual
```

**Comment by jun0s** [comment]
what do u think? 

$ ./john
Sorry, SSSE3 is required for this build

**Comment by magnumripper** [comment]
%@&§*#

**Comment by jun0s** [comment]
l know!! sorry :)

**Comment by magnumripper** [comment]
Don't get me wrong, it's really good we iron this out before Jumbo-2 release!

I just rechecked; it works for me, and it looks right in the debugger. An SSSE3 build does test for SSSE3 properly. Did you `make clean` and rebuild all of it?

**Comment by jun0s** [comment]
let me try again..

**Comment by jun0s** [comment]
i get download a new copy from github and nothing... 

$ ./john.exe
Sorry, SSSE3 is required for this build

configure: Testing build host's native CPU features
checking for Hyperthreading... yes
checking for MMX... yes
checking for SSE2... yes
checking for SSSE3... yes
checking for SSE4.1... no

some of the context... Windows 8.1 x64 and cygwin x32 updated....

**Comment by magnumripper** [comment]
I have no idea what is wrong then. You have a couple of alternatives though: First, you can force that exe you already built to run, like this, and see if it works:
```
CPUID_DISABLE=1 ./john (...)
```
Theoretically that should crash with a segfault though - unless our CPU detection lies.

Or, you could build it without SSSE3 support and see if that helps (SSSE3 doesn't make much difference anyway)
```
./configure CPPFLAGS=-mno-ssse3 && make -s clean && make -sj4
```

Also, why don't you run cygwin64? There's a significant performance boost with some formats.

**Comment by jun0s** [comment]
Works... 
$ CPUID_DISABLE=1 ./john.exe
John the Ripper 1.8.0.6-jumbo-1-bleeding OMP [cygwin 32-bit SSSE3-ac]

maybe try using cygwin64... i need some 32bits programs, because can run on x64 and x32 systems,  u can be specific about which formats? and thank u very much!!

**Comment by magnumripper** [comment]
> Works... 

Does it pass a full `CPUID_DISABLE=1 ./john.exe --test=0` run too?

> u can be specific about which formats?

Most, really. For example, the NT format is something like 50% faster using 64-bit.

**Comment by jun0s** [comment]
nope.. 

fopen: $JOHN/john.ini: No such file or directory 

sound good... thanks.

**Comment by magnumripper** [comment]
> fopen: $JOHN/john.ini: No such file or directory

That's what you get if you move john.exe away from the "run" directory without building 'systemwide'. It needs all files in "run".

**Comment by jun0s** [comment]
same on cygwin x64
$ file john.exe
john.exe: PE32+ executable (console) x86-64 (stripped to external PDB), for MS Windows

$ ./john
Sorry, SSSE3 is required for this build

... i have all the files in the run directory, what do u think i have wrong? sorry for my bad and stupid english... lol

**Comment by jun0s** [comment]
well, maybe i need to disable SSSE3 like u say...

**Comment by magnumripper** [comment]
I once again checked a 32-bit (cross-compile) SSSE3 build and verified (stepped) the assembler code. All is correct. @jfoug would you be able to do the same with a 'real' cygwin32 SSSE3 build? Basically I did it like this
```
$ gdb --args ../run/john
(gdb) b CPU_detect
Breakpoint 1 at 0x83e9079
(gdb) r
Starting program: ../run/john 
[Thread debugging using libthread_db enabled]

Breakpoint 1, 0x083e9079 in CPU_detect ()
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.149.el6_6.7.i686 keyutils-libs-1.4-5.el6.i686 krb5-libs-1.10.3-37.el6_6.i686 libcom_err-1.41.12-21.el6.i686 libselinux-2.0.94-5.8.el6.i686 nss-softokn-freebl-3.14.3-22.el6_6.i686 openssl-1.0.1e-30.el6_6.11.i686 zlib-1.2.3-29.el6.i686
(gdb) disas
Dump of assembler code for function CPU_detect:
=> 0x083e9079 <+0>:	push   %ebx
   0x083e907a <+1>:	pushf  
   0x083e907b <+2>:	pushf  
   0x083e907c <+3>:	xorl   $0x200000,(%esp)
   0x083e9083 <+10>:	popf   
   0x083e9084 <+11>:	pushf  
   0x083e9085 <+12>:	pop    %eax
   0x083e9086 <+13>:	xor    (%esp),%eax
   0x083e9089 <+16>:	popf   
   0x083e908a <+17>:	and    $0x200000,%eax
   0x083e908f <+22>:	je     0x83e9101 <CPU_detect_ret>
   0x083e9091 <+24>:	xor    %eax,%eax
   0x083e9093 <+26>:	cpuid  
   0x083e9095 <+28>:	test   %eax,%eax
   0x083e9097 <+30>:	je     0x83e9101 <CPU_detect_ret>
   0x083e9099 <+32>:	push   %ecx
   0x083e909a <+33>:	mov    $0x1,%eax
   0x083e909f <+38>:	cpuid  
   0x083e90a1 <+40>:	and    $0x4000200,%ecx
   0x083e90a7 <+46>:	cmp    $0x4000200,%ecx
   0x083e90ad <+52>:	je     0x83e90b4 <CPU_detect_AVX2>
   0x083e90af <+54>:	pop    %ecx
   0x083e90b0 <+55>:	xor    %eax,%eax
   0x083e90b2 <+57>:	jmp    0x83e9101 <CPU_detect_ret>
End of assembler dump.
(gpu) si
0x083e907a in CPU_detect ()
(gpu) si
       (...and so on...)
0x083e90b4 in CPU_detect_AVX2 ()
(gdb) disas
Dump of assembler code for function CPU_detect_AVX2:
=> 0x083e90b4 <+0>:	mov    $0x1,%eax
   0x083e90b9 <+5>:	cpuid  
   0x083e90bb <+7>:	pop    %ecx
   0x083e90bc <+8>:	xchg   %eax,%edx
   0x083e90bd <+9>:	and    $0x4000000,%eax
   0x083e90c2 <+14>:	je     0x83e9101 <CPU_detect_ret>
   0x083e90c4 <+16>:	xchg   %eax,%edx
   0x083e90c5 <+17>:	and    $0xf0ff0,%eax
   0x083e90ca <+22>:	or     $0x1,%al
   0x083e90cc <+24>:	cmp    $0x444d4163,%ecx
   0x083e90d2 <+30>:	je     0x83e90f7 <CPU_detect_AMD>
   0x083e90d4 <+32>:	cmp    $0x736c7561,%ecx
   0x083e90da <+38>:	je     0x83e9101 <CPU_detect_ret>
   0x083e90dc <+40>:	cmp    $0x6c65746e,%ecx
   0x083e90e2 <+46>:	jne    0x83e9101 <CPU_detect_ret>
   0x083e90e4 <+48>:	cmp    $0x5,%ah
   0x083e90e7 <+51>:	je     0x83e9101 <CPU_detect_ret>
   0x083e90e9 <+53>:	cmp    $0xf,%ah
   0x083e90ec <+56>:	je     0x83e9101 <CPU_detect_ret>
   0x083e90ee <+58>:	cmp    $0x106a1,%eax
   0x083e90f3 <+63>:	je     0x83e9101 <CPU_detect_ret>
   0x083e90f5 <+65>:	jmp    0x83e9101 <CPU_detect_ret>
End of assembler dump.
```
The 0x4000200 is correct for SSE2 + SSSE3 and I also stepped through all of the code until return. The "AVX2" seen is just a somewhat misnamed label.

**Comment by jfoug** [comment]
```
$ ./configure CPPFLAGS=-mno-sse4
......
Configured for building John the Ripper jumbo:

Target CPU ................................. i686 SSSE3, 32-bit LE
AES-NI support ............................. run-time detection
Target OS .................................. cygwin
Cross compiling ............................ no
Legacy arch header ......................... x86-sse.h

Optional libraries/features found:
Fuzzing test ............................... no
Experimental code .......................... no
OpenMPI support (default disabled) ......... no
Fork support ............................... yes
OpenMP support ............................. yes (not for fast formats)
CUDA support (default disabled by OpenCL) .. no
OpenCL support ............................. yes
Generic crypt(3) format .................... yes
Rexgen (extra cracking mode) ............... no
GMP (PRINCE mode and faster SRP formats) ... yes
PCAP (vncpcap2john and SIPdump) ............ yes
Z (pkzip format, gpg2john) ................. yes
BZ2 (gpg2john extra decompression logic) ... yes
128-bit integer (faster PRINCE mode) ....... no
Memory map (share/page large files) ........ yes

Development options (these may hurt performance when enabled):
Memdbg memory debugging settings ........... disabled
AddressSanitizer ("ASan") .................. disabled
UndefinedBehaviorSanitizer ("UbSan") ....... disabled

Install missing libraries to get any needed features that were omitted.

Configure finished.  Now 'make clean && make -s' to compile.


$ make -s clean && make -sj8

```

```
$ gdb --args ../run/john
GNU gdb (GDB) 7.8
Copyright (C) 2014 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "i686-pc-cygwin".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ../run/john...done.
(gdb) b CPU_detect
Breakpoint 1 at 0x757079
(gdb) r
Starting program: /cygdrive/c/projects/phpbb/johnripper/john-1.7.9/bleed32/run/john
[New Thread 20184.0x3808]
[New Thread 20184.0x4fc4]

Breakpoint 1, 0x00757079 in CPU_detect ()
(gdb) disas
Dump of assembler code for function CPU_detect:
=> 0x00757079 <+0>:     push   %ebx
   0x0075707a <+1>:     pushf
   0x0075707b <+2>:     pushf
   0x0075707c <+3>:     xorl   $0x200000,(%esp)
   0x00757083 <+10>:    popf
   0x00757084 <+11>:    pushf
   0x00757085 <+12>:    pop    %eax
   0x00757086 <+13>:    xor    (%esp),%eax
   0x00757089 <+16>:    popf
   0x0075708a <+17>:    and    $0x200000,%eax
   0x0075708f <+22>:    je     0x75712b <CPU_detect_ret>
   0x00757095 <+28>:    xor    %eax,%eax
   0x00757097 <+30>:    cpuid
   0x00757099 <+32>:    test   %eax,%eax
   0x0075709b <+34>:    je     0x75712b <CPU_detect_ret>
   0x007570a1 <+40>:    push   %ecx
   0x007570a2 <+41>:    mov    $0x1,%eax
   0x007570a7 <+46>:    cpuid
   0x007570a9 <+48>:    and    $0x4080200,%ecx
   0x007570af <+54>:    cmp    $0x4080200,%ecx
   0x007570b5 <+60>:    je     0x7570bc <CPU_detect_AVX2>
   0x007570b7 <+62>:    pop    %ecx
   0x007570b8 <+63>:    xor    %eax,%eax
   0x007570ba <+65>:    jmp    0x75712b <CPU_detect_ret>
End of assembler dump.
(gdb) si
0x0075707a in CPU_detect ()
(gdb) si
0x0075707b in CPU_detect ()
(gdb) si
0x0075707c in CPU_detect ()
(gdb) si
0x00757083 in CPU_detect ()
(gdb) si
0x00757084 in CPU_detect ()
(gdb) si
0x00757085 in CPU_detect ()
(gdb) si
0x00757086 in CPU_detect ()
(gdb) si
0x00757089 in CPU_detect ()
(gdb) si
0x0075708a in CPU_detect ()
(gdb) si
0x0075708f in CPU_detect ()
(gdb) si
0x00757095 in CPU_detect ()
(gdb) si
0x00757097 in CPU_detect ()
(gdb) si
0x00757099 in CPU_detect ()
(gdb) si
0x0075709b in CPU_detect ()
(gdb) si
0x007570a1 in CPU_detect ()
(gdb) si
0x007570a2 in CPU_detect ()
(gdb) si
0x007570a7 in CPU_detect ()
(gdb) si
0x007570a9 in CPU_detect ()
(gdb) si
0x007570af in CPU_detect ()
(gdb) si
0x007570b5 in CPU_detect ()
(gdb) si
0x007570bc in CPU_detect_AVX2 ()
(gdb) disas
Dump of assembler code for function CPU_detect_AVX2:
=> 0x007570bc <+0>:     mov    $0x1,%eax
   0x007570c1 <+5>:     cpuid
   0x007570c3 <+7>:     and    $0x1c000000,%ecx
   0x007570c9 <+13>:    cmp    $0x1c000000,%ecx
   0x007570cf <+19>:    jne    0x7570e0 <CPU_detect_pop_ecx_fail>
   0x007570d1 <+21>:    xor    %ecx,%ecx
   0x007570d3 <+23>:    push   %eax
   0x007570d4 <+24>:    push   %edx
   0x007570d5 <+25>:    xgetbv
   0x007570d8 <+28>:    and    $0x6,%al
   0x007570da <+30>:    cmp    $0x6,%al
   0x007570dc <+32>:    pop    %edx
   0x007570dd <+33>:    pop    %eax
   0x007570de <+34>:    je     0x7570e5 <CPU_detect_AVX>
End of assembler dump.
(gdb)
```

I think I get the same as you.  The je jump is taken, not the return 0

**Comment by jfoug** [comment]
I really think this has to do again with testing for things which for whatever reason, are NOT in the silicone of this CPU and we are testing for them, BUT which are not actually currently needed.

**Comment by jfoug** [comment]
The easiest way to tell, is to create 'simple' C code that checks every single item being looked for, and have it output when one of the options is NOT found.  That is how I found what was lacking for the AVX2 VM stuff.

**Comment by magnumripper** [comment]
> I think I get the same as you.

No, you have the bug! Excellent, then you can nail it. You have this:
```
   0x007570a9 <+48>:    and    $0x4080200,%ecx
   0x007570af <+54>:    cmp    $0x4080200,%ecx
```
That's SSE2 + SSE4.1 + SSSE3, so it tests too much. I had this:
```
   0x083e90a1 <+40>:    and    $0x4000200,%ecx
   0x083e90a7 <+46>:    cmp    $0x4000200,%ecx
```
That's SSE2 + SSSE3, which is correct.

The question is WHY it ends up incorrect? And why is it correct for me? Try to follow the exact macros (both `JOHN_SSSE3` and `__SSSE3__` type macros, and `CPU_REQ_xxx`) and see what ends up wrong. Also, try "rm x86.o" and `make` without `-s` so you see the build parameters for x86.S

**Comment by magnumripper** [comment]
Your result may be caused by not using `--disable-native-tests` (I did). I'm not sure CPPFLAGS is supplied when compiling asm files so you probably end up with the `-march=native` having free hands.

Unfortunately this may mean it's not the same problem @jun0s have.

**Comment by jfoug** [comment]
I will re-build. I did not do the --disable-native-tests, so I bet ```__SSE41__``` was defined.

**Comment by magnumripper** [comment]
BTW I did verify today that the SSSE3 test we do in ./configure is correct. It is. I was thinking we could be using some instruction that happen to also be in some half-way AMD instruction set like SSE4a and that this would explain @jun0s' problems (resulting in AC choosing SSSE3 when it should really be SSE2). But that's not it.

**Comment by magnumripper** [comment]
> I really think this has to do again with testing for things which for whatever reason, are NOT in the silicone of this CPU and we are testing for them, BUT which are not actually currently needed.

Nope. We test for exactly the right thing, there's no doubt at all. SSSE3 and SSE4.x are much easier to test for than AVX and AVX2. They are single flags in the original leaf 1 function.

**Comment by jfoug** [comment]
Could he have some AMD/Cyrix chip that has some differences such as supporting 'some' of the SSSE3 instructions, but not all, and thus not setting the bit listing that SSSE3 is valid ???

We really need to find out exactly WHAT silicon he has in his PC/laptop.

**Comment by magnumripper** [comment]
As far as I could see on wikipedia, there's no CPU made that would have PSHUFB but not full SSSE3. But who knows.

**Comment by jun0s** [comment]
sorry guys... btw i have intel E4500, very old processor but works just fine.. and i am not a programmer but i understand some stuffs.. i do ur tests here...

(gdb) si
0x0063c5c7 in CPU_detect_or_fallback (argv=0x28cc5c, make_check=0) at john.c:1358
1358            if (!CPU_detect()) {
(gdb) disa
(gdb) disa
(gdb) si
0x0063c5c9      1358            if (!CPU_detect()) {
(gdb) disa
(gdb) disas
Dump of assembler code for function CPU_detect_or_fallback:
   0x0063c5ad <+0>:     sub    $0x1c,%esp
   0x0063c5b0 <+3>:     movl   $0x7d9a8f,(%esp)
   0x0063c5b7 <+10>:    call   0x6a131c <getenv>
   0x0063c5bc <+15>:    test   %eax,%eax
   0x0063c5be <+17>:    je     0x63c5c2 <CPU_detect_or_fallback+21>
   0x0063c5c0 <+19>:    jmp    0x63c612 <CPU_detect_or_fallback+101>
   0x0063c5c2 <+21>:    call   0x676079 <CPU_detect>
   0x0063c5c7 <+26>:    test   %eax,%eax
=> 0x0063c5c9 <+28>:    jne    0x63c612 <CPU_detect_or_fallback+101>
   0x0063c5cb <+30>:    call   0x6a11dc <__getreent>
   0x0063c5d0 <+35>:    mov    0xc(%eax),%eax
   0x0063c5d3 <+38>:    movl   $0x7d9a9d,0x8(%esp)
   0x0063c5db <+46>:    movl   $0x7d9aa4,0x4(%esp)
   0x0063c5e3 <+54>:    mov    %eax,(%esp)
   0x0063c5e6 <+57>:    call   0x6a12ac <fprintf>
   0x0063c5eb <+62>:    cmpl   $0x0,0x24(%esp)
   0x0063c5f0 <+67>:    je     0x63c5fe <CPU_detect_or_fallback+81>
   0x0063c5f2 <+69>:    movl   $0x0,(%esp)
   0x0063c5f9 <+76>:    call   0x6a1254 <exit>
   0x0063c5fe <+81>:    movl   $0x566,0x4(%esp)
   0x0063c606 <+89>:    movl   $0x7d913a,(%esp)
   0x0063c60d <+96>:    call   0x64dc60 <real_error>
---Type <return> to continue, or q <return> to quit---
   0x0063c612 <+101>:   add    $0x1c,%esp
   0x0063c615 <+104>:   ret
End of assembler dump.
(gdb) si
1378                    fprintf(stderr, "Sorry, %s is required for this build\n",
(gdb) disas
Dump of assembler code for function CPU_detect_or_fallback:
   0x0063c5ad <+0>:     sub    $0x1c,%esp
   0x0063c5b0 <+3>:     movl   $0x7d9a8f,(%esp)
   0x0063c5b7 <+10>:    call   0x6a131c <getenv>
   0x0063c5bc <+15>:    test   %eax,%eax
   0x0063c5be <+17>:    je     0x63c5c2 <CPU_detect_or_fallback+21>
   0x0063c5c0 <+19>:    jmp    0x63c612 <CPU_detect_or_fallback+101>
   0x0063c5c2 <+21>:    call   0x676079 <CPU_detect>
   0x0063c5c7 <+26>:    test   %eax,%eax
   0x0063c5c9 <+28>:    jne    0x63c612 <CPU_detect_or_fallback+101>
=> 0x0063c5cb <+30>:    call   0x6a11dc <__getreent>
   0x0063c5d0 <+35>:    mov    0xc(%eax),%eax
   0x0063c5d3 <+38>:    movl   $0x7d9a9d,0x8(%esp)
   0x0063c5db <+46>:    movl   $0x7d9aa4,0x4(%esp)
   0x0063c5e3 <+54>:    mov    %eax,(%esp)
   0x0063c5e6 <+57>:    call   0x6a12ac <fprintf>
   0x0063c5eb <+62>:    cmpl   $0x0,0x24(%esp)
   0x0063c5f0 <+67>:    je     0x63c5fe <CPU_detect_or_fallback+81>
   0x0063c5f2 <+69>:    movl   $0x0,(%esp)
   0x0063c5f9 <+76>:    call   0x6a1254 <exit>
   0x0063c5fe <+81>:    movl   $0x566,0x4(%esp)
   0x0063c606 <+89>:    movl   $0x7d913a,(%esp)
   0x0063c60d <+96>:    call   0x64dc60 <real_error>
---Type <return> to continue, or q <return> to quit---
   0x0063c612 <+101>:   add    $0x1c,%esp
   0x0063c615 <+104>:   ret
End of assembler dump.
(gdb)

maybe i can help u with some other tests?

**Comment by magnumripper** [comment]
OK that's the 64-bit build. But you disassembled the wrong section. Try `disas CPU_detect`

**Comment by magnumripper** [comment]
The E4500 has SSSE3 so autoconf is correct.

**Comment by jun0s** [comment]
its x32 build... 

(gdb) disas CPU_detect
Dump of assembler code for function CPU_detect:
   0x00676079 <+0>:     push   %ebx
   0x0067607a <+1>:     pushf
   0x0067607b <+2>:     pushf
   0x0067607c <+3>:     xorl   $0x200000,(%esp)
   0x00676083 <+10>:    popf
   0x00676084 <+11>:    pushf
   0x00676085 <+12>:    pop    %eax
   0x00676086 <+13>:    xor    (%esp),%eax
   0x00676089 <+16>:    popf
   0x0067608a <+17>:    and    $0x200000,%eax
   0x0067608f <+22>:    je     0x676101 <CPU_detect_ret>
   0x00676091 <+24>:    xor    %eax,%eax
   0x00676093 <+26>:    cpuid
   0x00676095 <+28>:    test   %eax,%eax
   0x00676097 <+30>:    je     0x676101 <CPU_detect_ret>
   0x00676099 <+32>:    push   %ecx
   0x0067609a <+33>:    mov    $0x1,%eax
   0x0067609f <+38>:    cpuid
   0x006760a1 <+40>:    and    $0x4000200,%ecx
   0x006760a7 <+46>:    cmp    $0x4000200,%ecx
   0x006760ad <+52>:    je     0x6760b4 <CPU_detect_AVX2>
   0x006760af <+54>:    pop    %ecx
---Type <return> to continue, or q <return> to quit---
   0x006760b0 <+55>:    xor    %eax,%eax
   0x006760b2 <+57>:    jmp    0x676101 <CPU_detect_ret>
End of assembler dump.
(gdb)

**Comment by magnumripper** [comment]
Interesting, this is the correct code. Yet somehow it fails the test of "SSE2 + SSSE3" (0x4000200).

**Comment by magnumripper** [comment]
Is there some kind of "cpuid" command in Cygwin? There is one for Linux.

**Comment by jun0s** [comment]
sorry... not in cygwin, and yes... E4500 have a SSSE3 instruction. I am looking for a windows equivalent..

**Comment by jfoug** [comment]
CPU-Z but it is not the same as cpuid for linux.

**Comment by magnumripper** [comment]
Try pasting this to a file "cpuid.c"
```
/*
 * This software is Copyright (c) 2015 magnum, and it is hereby released
 * to the general public under the following terms:  Redistribution and use in
 * source and binary forms, with or without modification, are permitted.
 */
#include <stdio.h>
#include <stdlib.h>
#if _MSC_VER
#include <intrin.h>
#endif

int usage(char *name)
{
	fprintf(stderr, "Usage: %s <leaf> [sub-leaf]\n", name);
	fprintf(stderr, "\nCalls cpuid with the given function id and displays the results.\n");

	return EXIT_FAILURE;
}

int main(int argc, char **argv)
{
	int i, regs[4];
	int str32[4];
	char *vendor = (char*)str32;
	unsigned int fn, sb;

	if (argc < 2 || argc > 3)
		exit(usage(argv[0]));

	fn = atoi(argv[1]);
	sb = (argc == 3) ? atoi(argv[2]) : 0;

#if _MSC_VER
	__cpuid(regs, 1);
#else
	__asm__ __volatile__(
#if __x86_64__
		"pushq %%rbx\n\t"
#else
		"pushl %%ebx\n\t"
#endif
		"cpuid\n\t" "movl %%ebx ,%[ebx]\n\t"
#if __x86_64__
		"popq %%rbx\n\t"
#else
		"popl %%ebx\n\t"
#endif
		:"=a"(regs[0]),[ebx] "=r"(regs[1]), "=c"(regs[2]), "=d"(regs[3])
		:"a"(fn), "c"(sb));
#endif
	printf("EAX=%xh, ECX=%xh CPUID\n", fn, sb);
	switch(fn) {

	case 0:
		str32[0] = regs[1];
		str32[1] = regs[3];
		str32[2] = regs[2];
		printf("Vendor ID: '%.12s' max leaf %xh\n", vendor, regs[0]);
		break;

	case 1:
	{
		int model = (regs[0] >> 4) & 15;
		int family = (regs[0] >> 8) & 15;
		int stepping = regs[0] & 15;
		int type = (regs[0] >> 12) & 3;
		int xmodel = (regs[0] >> 16) & 15;
		int xfamily = (regs[0] >> 20) & 255;

		if (family == 15)
			family += xfamily;
		if (family == 6 || family == 15)
			model += (xmodel << 4);
		printf("Model %d, family %d, stepping %d, type %d\n",
		       model, family, stepping, type);
		puts("                31                             0");
		for (i = 1; i < 4; i++) {
			int j;

			printf("E%cX:  %08x  ", 'A' + i, regs[i]);
			for (j = 31; j >= 0; j--)
				printf("%c", regs[i] & (1 << j) ? '1' : '0');
			printf("\n");
		}
	}
	break;

	case 3:
		puts("                31                             0");
		for (i = 2; i < 4; i++) {
			int j;

			printf("E%cX:  %08x  ", 'A' + i, regs[i]);
			for (j = 31; j >= 0; j--)
				printf("%c", regs[i] & (1 << j) ? '1' : '0');
			printf("\n");
		}
		break;

	case 7:
		puts("                31                             0");
		for (i = 1; i < 3; i++) {
			int j;

			printf("E%cX:  %08x  ", 'A' + i, regs[i]);
			for (j = 31; j >= 0; j--)
				printf("%c", regs[i] & (1 << j) ? '1' : '0');
			printf("\n");
		}
		break;

	case 0x80000000:
		printf("Max extended leaf: %xh\n", regs[0]);
		break;

	case 0x80000002:
	case 0x80000003:
	case 0x80000004:
		for (i = 0; i < 4; i++)
			str32[i] = regs[i];
		printf("Vendor brand[%d]: '%.16s'\n", (fn & 7) - 2, vendor);
		break;

	default:
		puts("                31                             0");
		for (i = 0; i < 4; i++) {
			int j;

			printf("E%cX:  %08x  ", 'A' + i, regs[i]);
			for (j = 31; j >= 0; j--)
				printf("%c", regs[i] & (1 << j) ? '1' : '0');
			printf("\n");
		}
	}

	exit(EXIT_SUCCESS);
}
```
Compile it with `gcc cpuid.c -o cpuid`. Then please post the output of `./cpuid 1`

**Comment by jun0s** [comment]
$ ./cpuid 1
EAX=1h, ECX=0h CPUID
Model 15, family 6, stepping 13, type 0
                31                             0
EBX:  0028cc9c  00000000001010001100110010011100
ECX:  0000e39d  00000000000000001110001110011101
EDX:  bfebfbff  10111111111010111111101111111111

**Comment by magnumripper** [comment]
SSSE3 is bit 9 of ECX... and that one is set. So what the heck is wrong then? Could it be some of the original non-Jumbo tests that fail?

**Comment by magnumripper** [comment]
I see the bug now. My bad, commit coming soon.

**Comment by magnumripper** [comment]
Try f2f8cb8

**Comment by magnumripper** [comment]
I confused EDX with ECX :sunglasses:

**Comment by jun0s** [comment]
Thats works for me.. THANK U!!

$ ./john.exe --restore:somesession
fopen: $JOHN/john.ini: No such file or directory

i have all the binary and scripts in the default directory "run"  but i do the configure with the prefix=/usr

**Comment by magnumripper** [comment]
Cool, I'm glad this is fixed.

> i have all the binary and scripts in the default directory "run" but i do the configure with the prefix=/usr

The prefix mechanism is subject to change, I recommend not using it. Use JOHN_SYSTEMWIDE as described in params.h and [here](https://github.com/magnumripper/JohnTheRipper/wiki/Fallback-binary-chains)

**Comment by jfoug** [comment]
I am not 100% sure that JOHN_SYSTEMWIDE will work for cygwin ??

**Comment by magnumripper** [comment]
Why wouldn't it? It merely affects where john looks for config files and so on.

**Comment by jfoug** [comment]
IIRC, there was some '#define' magic that excluded cygwin (exe differences, path character differences or something).  Since I am not looking at code right now, but dredging up several year old memories, I may simply be talking out my arse ;)

**Comment by magnumripper** [comment]
I think you confuse it with CPU_FALLBACK. Apparently that (the `execv(fallback_binary, argv)`) is disabled for Cygwin and djgpp, but the basic "Sorry, AVX2 is required" is not (as seen already).
